src/Layer/LabelLayer.js

```diff
@@ -0,0 +1,189 @@
+                coord.as(this.crs, coord);
+                coord.toVector3(label.position);
+                label.needsToBeMoved = true;
+
+                if (f.size == 2) {
+                    label.needsAltitude = true;
+                }
+
+                labels.push(label);
+            });
+        });
+
+        return labels;
+    }
+
+    update(context, layer, node, parent) {
+        if (!parent && node.children.length) {
+            // if node has been removed dispose three.js resource
+            ObjectRemovalHelper.removeChildrenAndCleanupRecursively(this, node);
+            return;
+        }
+
+        if (this.frozen || !node.visible) {
+            return;
+        }
+
+        if (node.layerUpdateState[this.id] === undefined) {
+            node.layerUpdateState[this.id] = new LayerUpdateState();
+        }
+
+        if (!node.layerUpdateState[this.id].canTryUpdate()) {
+            return;
+        }
+
+        const extentsDestination = node.getExtentsByProjection(this.source.projection) || [node.extent];
+
+        const extentsSource = [];
+        for (const extentDest of extentsDestination) {
+            const ext = this.source.projection == extentDest.crs ? extentDest : extentDest.as(this.source.projection);
+            if (extentInsideSource(ext, this.source)) {
+                node.layerUpdateState[this.id].noMoreUpdatePossible();
+                return;
+            }
+            extentsSource.push(extentDest);
+        }
+        node.layerUpdateState[this.id].newTry();
+
+        const command = {
+            layer: this,
+            extentsSource,
+            view: context.view,
+            threejsLayer: this.threejsLayer,
+            requester: node,
+        };
+
+        return context.scheduler.execute(command).then((result) => {
+            if (!result) { return; }
+
+            result.forEach((labels) => {
+                if (!node.parent) {
+                    labels.forEach(l => ObjectRemovalHelper.removeChildrenAndCleanupRecursively(this, l));
+                    return;
+                }
+
+                labels.forEach((label) => {
+                    if (label.parent) { return; }
+
+                    label.minlevel = node.extent.zoom;
+
+                    if (label.needsToBeMoved) {
+                        node.worldToLocal(label.position);
+                        label.needsToBeMoved = false;
+                    }
+
+                    node.add(label);
+                    context.view.mainLoop.gfxEngine.label2dRenderer.domElement.appendChild(label.content);
+
+                    if (label.needsAltitude) {
+                        // TODO: add a Style property for the offset
+                        label.position.z = DEMUtils.getElevationValueAt(this.parent, label.coordinates) + 2;
+                    }
+
+                    label.visible = true;
+                    label.updateDimensions();
+                    label.updateMatrixWorld();
```

Comment:
May be override `Label#updateMatrixWorld` to encapsulate `updateDimensions`?

Reply:
Won't `updateDimensions` be called too many times if I do this ?